### PR TITLE
feat (metrics): #1595 implement undesired event pings for loader and missing image

### LINKED ISSUE #1595: Track metrics for "bad" app states
It would be good to know things like:
- people hitting the "hang on tight", and for how long
- missing metadata/images in highlights

### PR DESCRIPTION
**Pull request by rlr**
This needs to wait for @ncloudioj to be back to r? the new pings and create the new table. Not sure if @sarracini or @k88hudson is around to r? the addon and frontend code changes?

Fixes #1595

### CODE REVIEW

File: common/action-manager.js [hunk 1]
```diff
@@ -203,6 +204,16 @@ function NotifyEvent(data) {
   return Notify("NOTIFY_USER_EVENT", data);
 }
 
+function NotifyUndesiredEvent(data) {
+  if (!eventConstants.defaultPage === data.page) {
+    throw new Error(`${data.page} is not a valid page`);
+  }
+  if (!eventConstants.undesiredEvents.has(data.event)) {
```
Comment (ncloudioj) [inline common/action-manager.js]: Nice little check here 👍

File: content-src/components/NewTabPage/NewTabPage.js [hunk 2]
```diff
@@ -38,6 +38,24 @@ const NewTabPage = React.createClass({
...
+      }));
+      this.loaderShownAt = Date.now();
+    }
+  },
+  componentDidUpdate(prevProps) {
+    if (this.props.isReady && this.loaderShownAt) {
+      this.props.dispatch(actions.NotifyUndesiredEvent({
+        event: "HIDE_LOADER",
+        page: PAGE_NAME,
+        timeShown: Date.now() - this.loaderShownAt
```
Comment (rlr) [inline content-src/components/NewTabPage/NewTabPage.js]: I tried this but webpack no like: 

```ERROR in ./common/AbsPerf.js
Module not found: Error: Cannot resolve module 'chrome' in /Users/rlr/dev/activity-streams/common
 @ ./common/AbsPerf.js 8:17-34
```

Comment (ncloudioj) [inline content-src/components/NewTabPage/NewTabPage.js]: Interesting.

Hey @dmose, iirc `AbsPerf.js` should work on the content side, correct?

File: data_dictionary.md [hunk 3]
```diff
@@ -73,6 +73,20 @@ The Activity Stream addon sends various types of pings to the backend (HTTPS POS
...
 ```
 
+# Example Activity Stream `undesired event` Log
+
+```json
+{
+  "action": "activity_stream_undesired_event",
+  "addon_version": "1.0.12",
+  "client_id": "26288a14-5cc4-d14f-ae0a-bb01ef45be9c",
+  "event": "HIGHLIGHT_WITHOUT_IMAGE",
```
Comment (ncloudioj) [inline data_dictionary.md]: s/\t/\s/, same as "page" and "tab_id" please.

**Comment by ncloudioj** [comment]
:shipit: I think we can revisit the AbsPerf thing later.

**Comment by rlr** [comment]
@ncloudioj ^^ I did the changes we discussed in IRC

File: addon/TabTracker.js [hunk 4]
```diff
@@ -112,6 +113,16 @@ TabTracker.prototype = {
     }
   },
 
+  handleUndesiredEvent(payload, experimentId) {
+    payload.action = "activity_stream_undesired_event";
+    payload.tab_id = tabs.activeTab.id;
+    this._setCommonProperties(payload, tabs.activeTab.url);
+    if (!("value" in payload)) {
```
Comment (ncloudioj) [inline addon/TabTracker.js]: So it doesn't support `"value" not in payload`? That sucks!

File: addon/TabTracker.js [hunk 5]
```diff
@@ -112,6 +113,16 @@ TabTracker.prototype = {
     }
   },
 
+  handleUndesiredEvent(payload, experimentId) {
+    payload.action = "activity_stream_undesired_event";
```
Comment (ncloudioj) [inline addon/TabTracker.js]: I personally hate those words like "undesired", "bad", "strange", how about "masga"? We could use this to track other app states next, not necessarily be "undesired". My apologies for my OCD, you can just rename this `action` here and leave others alone, as this is the only thing I'd see 😺

File: addon/TabTracker.js (hunk 4 above)
Comment (rlr) [inline addon/TabTracker.js]: heh! this isn't python 😡

**Pull request merged**